ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -265,45 +294,104 @@ void checkRateLimit() {
+        final KsqlNode node = entry.getKey();
+        final PullQueryContext pullQueryContext = contextFactory.apply(entry.getValue());
+
+        futures.put(node, executorService.submit(() ->  {
+          final TableRows rows = routeQuery.routeQuery(
+              node, statement, executionContext, serviceContext, pullQueryContext);
+          final Optional<List<KsqlNode>> debugNodes = Optional.ofNullable(
+              routingOptions.isDebugRequest()
+                  ? Collections.nCopies(rows.getRows().size(), node) : null);
+          return new PullQueryResult(rows, debugNodes);
+        }));
+      }
+
+      final ImmutableList.Builder<KsqlLocation> nextRoundRemaining = ImmutableList.builder();
+      for (Map.Entry<KsqlNode, Future<PullQueryResult>>  entry : futures.entrySet()) {
+        final Future<PullQueryResult> future = entry.getValue();
+        final KsqlNode node = entry.getKey();
+        try {
+          final PullQueryResult result = future.get();
+          result.getSourceNodes().ifPresent(sourceNodes::addAll);
+          schemas.add(result.getTableRows().getSchema());
+          tableRows.addAll(result.getTableRows().getRows());
+        } catch (ExecutionException e) {
+          LOG.debug("Error routing query {} to host {} at timestamp {} with exception {}",
+              statement.getStatementText(), node, System.currentTimeMillis(), e.getCause());
+          nextRoundRemaining.addAll(groupedByHost.get(node));
+        }
+      }
+      remainingLocations = nextRoundRemaining.build();
+
+      if (remainingLocations.size() == 0) {
+        validateSchemas(schemas);
         return new PullQueryResult(
-            routeQuery(node, statement, executionContext, serviceContext, pullQueryContext),
-            debugNode);
-      } catch (Exception t) {
-        LOG.debug("Error routing query {} to host {} at timestamp {} with exception {}",
-                  statement.getStatementText(), node, System.currentTimeMillis(), t);
+            new TableRows(statement.getStatementText(), queryId, Iterables.getLast(schemas),
+                tableRows),
+            sourceNodes.isEmpty() ? Optional.empty() : Optional.of(sourceNodes));
       }
     }
-    throw new MaterializationException(String.format(
-        "Unable to execute pull query: %s", statement.getStatementText()));
+  }
+
+  private static Map<KsqlNode, List<KsqlLocation>> groupByHost(
+      final ConfiguredStatement<Query> statement,
+      final List<KsqlLocation> locations,
+      final int round) {
+    final Map<KsqlNode, List<KsqlLocation>> groupedByHost = new LinkedHashMap<>();
+    for (KsqlLocation location : locations) {
+      // If one of the partitions required is out of nodes, then we cannot continue.
+      if (round >= location.getNodes().size()) {
+        throw new MaterializationException(String.format(
+            "Unable to execute pull query: %s", statement.getStatementText()));
```

Comment:
Consider adding a more descriptive error message so that we know at which point in the code the query failed and why.